### PR TITLE
Update rand version

### LINKED ISSUE #57724: rustc is using three versions of rand
I noticed this scrolling by while "Building stage0 compiler artifacts" today:
```rust
   Compiling rand_pcg v0.1.1
   Compiling rand v0.6.1
   Compiling flate2 v1.0.6
   Compiling rand v0.4.3
   Compiling rand v0.5.5
   Compiling winapi-util v0.1.1
```

Optimistically flagging E-easy

### PR DESCRIPTION
**Pull request by JohnTitor**
cc: #57724

r? @scottmcm

### CODE REVIEW

**Comment by JohnTitor** [comment]
@scottmcm Can you review this, or should I ask someone to do instead of you?

**Comment by mati865** [comment]
I don't think it fixes mentioned issue. It upgrades dependency for one package but Rust will still build 3 versions of rand: https://github.com/rust-lang/rust/blob/939a2639742f161811b500bc45bd06a1d63e9d18/Cargo.lock#L4179-L4181

**Comment by lzutao** [comment]
Right, the first one is 
https://github.com/rust-lang/rust/blob/939a2639742f161811b500bc45bd06a1d63e9d18/Cargo.lock#L1777-L1782
https://github.com/rust-lang/rust/blob/939a2639742f161811b500bc45bd06a1d63e9d18/Cargo.lock#L1885-L1890

**Comment by JohnTitor** [comment]
Well, this issue is about "Building stage0 compiler artifacts", I think. I checked build log `python x.py build --stage=0`, I saw `rand v0.6.1` only. Anything I'm missing?

**Comment by mati865** [comment]
The title "rustc is using three versions of rand" remains true when this PR is merged. I'm not reviewer but IMO this PR should only `cc` it.

**Comment by JohnTitor** [comment]
@mati865 Okay, I got it! I changed description following your opinion.

**Comment by mati865** [comment]
I don't see other problems here.

For the further work:
There is another crate in this repository which depends on `rand 0.5`: https://github.com/rust-lang/rust/blob/8e2063d02062ee9f088274690a97826333847e17/src/tools/rustc-workspace-hack/Cargo.toml#L55

I have no idea if upgrading it won't break tools build.
It's not possible to drop `rand 0.5` entirely just yet because `cargo` depends on `proptest 0.8.7` and cannot upgrade to `proptest 0.9` because test with `minimal-versions` fails.

It should be however possible to get rid of `rand 0.4` (in another PR) by updating few packages. I haven't ran tests but it got removed from `Cargo.lock` without obvious issues.
Do you want to investigate it?

**Comment by scottmcm** [comment]
Thanks, @JohnTitor!

@bors r+ rollup 

FWIW, I'd be happy for #57724 to close if `rustc`+`std` is down to one version, without worrying too much about whether other tools in the repo happen to use more.

**Comment by kennytm** [comment]
@bors rollup-

This PR changes Cargo.lock.

File: Cargo.lock [hunk 1]
```diff
@@ -2747,7 +2747,7 @@ version = "0.0.0"
 dependencies = [
  "graphviz 0.0.0",
  "log 0.4.6 (registry+https://github.com/rust-lang/crates.io-index)",
- "rand 0.5.5 (registry+https://github.com/rust-lang/crates.io-index)",
+ "rand 0.6.1 (registry+https://github.com/rust-lang/crates.io-index)",
```
Comment (jasonbking) [inline Cargo.lock]: Any chance of going to 0.6.5 instead? IIUC, that should bring in rand_os 0.1.3, which fixes SIGSYS crashes on illumos.

Comment (JohnTitor) [reply Cargo.lock]: I cannot decide that on my own, what do you think @scottmcm? (I updated from 0.6.1 to 0.6.5 on local, but the build was failed.)

Comment (mati865) [reply Cargo.lock]: @JohnTitor do you have error message from that build?

Comment (JohnTitor) [reply Cargo.lock]: @mati865 Yes, the message is here:
```
error[E0463]: can't find crate for `core`=============>                ] 16/22: libc
   --> C:\Users\username\.cargo\registry\src\github.com-1ecc6299db9ec823\libc-0.2.46\src\lib.rs:161:1
    |
161 | extern crate std as core;
    | ^^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `libc`.
```
I updated rand 0.6.1 by `cargo update -p rand:0.6.1` and changed [this line](https://github.com/rust-lang/rust/blob/7c19e1eed5418b8c02be65d678417b241ee01a3e/src/libstd/Cargo.toml#L29) to 0.6.

Comment (mati865) [reply Cargo.lock]: Looks like `std` failed to build, either it's issue with cache/incremental compilation or something really bad happened.

**Pull request merged**